### PR TITLE
Add the possibility to print a mail

### PR DESCRIPTION
**Pull request by tevincent**
https://github.com/Infomaniak/android-kMail/assets/149579879/21327e50-546f-4f61-bcd6-97dd9923c5f4

### CODE REVIEW

File: app/src/main/java/com/infomaniak/mail/utils/MessageBodyUtils.kt [hunk 1]
```diff
@@ -91,6 +97,74 @@ object MessageBodyUtils {
...
+            .attr("style", "margin-bottom: 40px; display: block")
+        message.cc.let { cc ->
+            messageDetailsDiv.insertPrintRecipientField(context.getString(R.string.ccTitle), *cc.toTypedArray())
+        }
+        message.to.let { to ->
+            messageDetailsDiv.insertPrintRecipientField(context.getString(R.string.toTitle), *to.toTypedArray())
+        }
+        message.sender?.let { sender ->
+            messageDetailsDiv.insertPrintRecipientField(context.getString(R.string.fromTitle), sender)
+        }
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/utils/MessageBodyUtils.kt:134]:
```suggestion
        val messageDetailsDiv = Element("div").attr("style", "margin-bottom: 40px; display: block")
        messageDetailsDiv.insertPrintRecipientField(context.getString(R.string.ccTitle), *message.cc.toTypedArray())
        messageDetailsDiv.insertPrintRecipientField(context.getString(R.string.toTitle), * message.to.toTypedArray())
        message.sender?.let { messageDetailsDiv.insertPrintRecipientField(context.getString(R.string.fromTitle), it) }
```

File: app/src/main/java/com/infomaniak/mail/utils/MessageBodyUtils.kt [hunk 2]
```diff
@@ -91,6 +97,74 @@ object MessageBodyUtils {
...
+    private fun Element.insertPrintRecipientField(prefix: String, vararg recipients: Recipient) {
+        if (recipients.isEmpty()) return
+
+        insertChildren(0,
+            insertField(prefix).appendText(
+                recipients.joinToString { recipient ->
+                    recipient.quotedDisplay()
+                }
+            )
+        )
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/utils/MessageBodyUtils.kt:155]:
```suggestion
        insertChildren(
            0,
            insertField(prefix).appendText(
                recipients.joinToString { it.quotedDisplay() },
            ),
        )
```

File: app/src/main/java/com/infomaniak/mail/utils/WebViewUtils.kt [hunk 3]
```diff
@@ -44,18 +45,28 @@ class WebViewUtils(context: Context) {
...
     private val jsBridgeScript by lazy { context.getJsBridgeScript() }
 
-    fun processHtmlForDisplay(html: String, isDisplayedInDarkMode: Boolean): String = with(HtmlFormatter(html)) {
-        addCommonDisplayContent(isDisplayedInDarkMode)
+    fun processHtmlForDisplay(
+        context: Context? = null,
+        html: String,
+        isDisplayedInDarkMode: Boolean,
+        message: Message? = null,
+        isForPrinting: Boolean = false
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/utils/WebViewUtils.kt:53]:
```suggestion
        isForPrinting: Boolean = false,
```

File: app/src/main/java/com/infomaniak/mail/utils/WebViewUtils.kt [hunk 4]
```diff
@@ -44,18 +45,28 @@ class WebViewUtils(context: Context) {
...
+        addCommonDisplayContent(isDisplayedInDarkMode, isForPrinting)
         return@with inject()
     }
 
-    fun processSignatureHtmlForDisplay(html: String, isDisplayedInDarkMode: Boolean): String = with(HtmlFormatter(html)) {
-        addCommonDisplayContent(isDisplayedInDarkMode)
+    fun processSignatureHtmlForDisplay(
+        html: String,
+        isDisplayedInDarkMode: Boolean,
+        isForPrinting: Boolean = false
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/utils/WebViewUtils.kt:62]:
```suggestion
        isForPrinting: Boolean = false,
```

File: app/src/main/res/layout/fragment_print_mail.xml [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        tools:visibility="gone">
+
+        <ProgressBar
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:layout_gravity="center" />
+
+    </FrameLayout>
+
+</FrameLayout>
```
Comment (KevinBoulongne) [inline app/src/main/res/layout/fragment_print_mail.xml:53]:
```suggestion
</FrameLayout>

```

File: app/src/main/res/raw/javascript_bridge.js [hunk 6]
```diff
@@ -23,3 +23,6 @@ function reportOverScroll(clientWidth, scrollWidth, messageUid) {
 function reportError(error, scriptFirstLine, messageUid) {
     window.kmail.reportError(error.name, error.message, error.stack, scriptFirstLine, messageUid);
 }
+function webviewFinishedLoading() {
```
Comment (KevinBoulongne) [inline app/src/main/res/raw/javascript_bridge.js:26]:
```suggestion

function webviewFinishedLoading() {
```

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/utils/MessageBodyUtils.kt [hunk 7]
```diff
@@ -91,6 +98,66 @@ object MessageBodyUtils {
...
+            0,
+            insertField(prefix).appendText(
+                recipients.joinToString { recipient -> recipient.quotedDisplay() }
+            ),
+        )
+    }
+
+    private fun Element.insertPrintDateField(prefix: String, date: String) {
+        insertChildren(0, insertField(prefix).appendText(date))
+    }
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/utils/MessageBodyUtils.kt:155]: The same way `insertPrintRecipientField` takes recipients as input, shouldn't `insertPrintDateField` take a Date object as input?

File: app/src/main/java/com/infomaniak/mail/utils/WebViewUtils.kt [hunk 8]
```diff
@@ -65,9 +76,15 @@ class WebViewUtils(context: Context) {
         registerScript(jsBridgeScript)
         registerBodyEncapsulation()
         registerBreakLongWords()
+        if (isForPrint) registerIsForPrint()
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/utils/WebViewUtils.kt:79]: This is not common between the message html and the signature html, this should not be grouped inside this method

File: app/src/main/java/com/infomaniak/mail/utils/WebViewUtils.kt [hunk 9]
```diff
@@ -100,7 +117,11 @@ class WebViewUtils(context: Context) {
 
     companion object {
         private const val DARK_BACKGROUND_STYLE_ID = "dark_background_style"
-        val jsBridge = JavascriptBridge() // TODO: Avoid excessive memory consumption with injection
+        var jsBridge: JavascriptBridge? = null // TODO: Avoid excessive memory consumption with injection
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/utils/WebViewUtils.kt:120]: Do we need to make it nullable if we're gonna init it anyway right after? Wouldn't `lateinit` be enough to avoid writting a line of code of `?.let` on a value that will always be set anyway?

File: app/src/main/java/com/infomaniak/mail/ui/newMessage/NewMessageFragment.kt
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/newMessage/NewMessageFragment.kt:1]: I think you don't need these imports anymore

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/MessageActionsBottomSheetDialog.kt [hunk 10]
```diff
@@ -17,11 +17,19 @@
...
+import android.os.ParcelFileDescriptor
+import android.print.PageRange
+import android.print.PrintAttributes
+import android.print.PrintDocumentAdapter
+import android.print.PrintManager
 import android.view.View
 import androidx.core.view.isVisible
 import androidx.navigation.fragment.navArgs
 import com.infomaniak.lib.core.utils.isNightModeEnabled
+import com.infomaniak.lib.core.utils.safeNavigate
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/MessageActionsBottomSheetDialog.kt:32]: You have many unused imports

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt [hunk 11]
```diff
@@ -599,6 +587,7 @@ class ThreadAdapter(
         val fullMessageWebViewClient get() = _fullMessageWebViewClient!!
 
         init {
+            if (isForPrinting) WebView.enableSlowWholeDocumentDraw()
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt:590]: Is this still needed now that you user the webview adapter and not the bitmap strategy?

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt [hunk 12]
```diff
@@ -57,22 +59,24 @@ import com.infomaniak.mail.utils.WebViewUtils.Companion.toggleWebViewTheme
...
+    onDeleteDraftClicked: ((message: Message) -> Unit)? = null,
+    onDraftClicked: ((message: Message) -> Unit)? = null,
+    onAttachmentClicked: ((attachment: Attachment) -> Unit)? = null,
+    onDownloadAllClicked: ((message: Message) -> Unit)? = null,
+    onReplyClicked: ((Message) -> Unit)? = null,
+    onMenuClicked: ((Message) -> Unit)? = null,
+    onAllExpandedMessagesLoaded: (() -> Unit)? = null,
+    navigateToNewMessageActivity: ((Uri) -> Unit)? = null,
+    navigateToAttendeeBottomSheet: ((List<Attendee>) -> Unit)? = null,
+    promptLink: ((String, ContextMenuType) -> Unit)? = null,
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt:79]: A probably better way than having to carry this around would be to replace all of these callbacks by a single instance of the ThreadAdapterCallbacks object. This list was growing out of hand anyway

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt [hunk 13]
```diff
@@ -104,6 +108,7 @@ class ThreadAdapter(
 
     init {
         threadAdapterCallbacks = ThreadAdapterCallbacks(
+            onBodyWebviewFinishedLoading,
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt:111]: It's usually spelled "WebView"

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt [hunk 14]
```diff
@@ -129,31 +134,37 @@ class ThreadAdapter(
...
     }
 
     override fun onBindViewHolder(holder: ThreadViewHolder, position: Int, payloads: MutableList<Any>) = runCatchingRealm {
         with(holder.binding) {
-            val payload = payloads.firstOrNull()
-            if (payload !is NotifyType) {
+            if (payloads.isEmpty()) {
                 super.onBindViewHolder(holder, position, payloads)
-                return
+                return@runCatchingRealm
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt:147]: Do we really need to specify this? I feel like it makes lines way heavier to read for no real gain, when you see `return` you already know the outer function will exit. It's not like like it could be ambiguous.

Same for others in here

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt [hunk 15]
```diff
@@ -129,31 +134,37 @@ class ThreadAdapter(
...
-            val message = messages[position]
-
-            when (payload) {
-                NotifyType.TOGGLE_LIGHT_MODE -> {
-                    isThemeTheSameMap[message.uid] = !isThemeTheSameMap[message.uid]!!
-                    holder.toggleContentAndQuoteTheme(message.uid)
+            payloads.forEach { payload ->
+                if (payload !is NotifyType) {
+                    super.onBindViewHolder(holder, position, payloads)
+                    return@runCatchingRealm
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt:152]: Wow, not sure we want to call super.onBindViewHolder multiple times, right?

Why did you need to change all of this part by the way?

File: app/src/main/java/com/infomaniak/mail/extensions/DateExtensions.kt [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+        // Adapts order of day, month and year according to the locale
+        val shortDateFormatter = DateTimeFormatter.ofLocalizedDate(FormatStyle.SHORT)
+        toInstant().atZone(ZoneId.systemDefault()).toLocalDate().format(shortDateFormatter)
+    } else {
+        // Fallback on unambiguous date format for any locale
+        format(FORMAT_DATE_CLEAR_MONTH_DAY_ONE_CHAR)
+    }
+}
+
+//TODO A refacto quand la PR de Gibran sera mergé
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/extensions/DateExtensions.kt:40]: Comments should be in english

File: app/src/main/java/com/infomaniak/mail/extensions/DateExtensions.kt
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/extensions/DateExtensions.kt:1]: Since I have a PR that creates an util, I think that you should not extract these functions to reduce merge conflicts. Leave the original function where it was and simply duplicate the code where you need to use it and modify it with yours needs and leave the TODO there

File: app/src/main/java/com/infomaniak/mail/extensions/DateExtensions.kt [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+fun Date.formatNumericalDayMonthYearWithTime(): String {
+    return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+        val shortDateFormatter = DateTimeFormatter.ofLocalizedDateTime(FormatStyle.SHORT)
+        toInstant().atZone(ZoneId.systemDefault()).format(shortDateFormatter)
+    } else {
+        // Fallback on unambiguous date format for any locale
+        format(FORMAT_EVENT_DATE)
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/extensions/DateExtensions.kt:47]: This is ambiguous. Use a date format where the month is spelled in all letters so there's no confusion for english speaking countries where 01/12 represents the 12th of January

File: app/src/main/java/com/infomaniak/mail/data/models/thread/Thread.kt [hunk 18]
```diff
@@ -203,17 +200,6 @@ class Thread : RealmObject {
...
-        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
-            // Adapts order of day, month and year according to the locale
-            val shortDateFormatter = DateTimeFormatter.ofLocalizedDate(FormatStyle.SHORT)
-            toInstant().atZone(ZoneId.systemDefault()).toLocalDate().format(shortDateFormatter)
-        } else {
-            // Fallback on unambiguous date format for any locale
-            format(FORMAT_DATE_CLEAR_MONTH_DAY_ONE_CHAR)
-        }
-    }
-
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/data/models/thread/Thread.kt:216]: Do not remove it from there, only duplicated what you need, this will help avoid useless merge conflicts

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import javax.inject.Inject
+
+@AndroidEntryPoint
+class PrintMailFragment : Fragment() {
+
+    @Inject
+    lateinit var localSettings: LocalSettings
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt:45]: We usually inject classes before the first method declaration and after the last variable declaration. This doesn't bother me that much but other in the team are more strict about this I think so I ended up doing as they told me for this

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+
+    private var _binding: FragmentPrintMailBinding? = null
+    private val binding get() = _binding!! // This property is only valid between onCreateView and onDestroyView
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt:48]: Can't you use the usual `by safeBinding()`?

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+
+    private val navigationArgs: PrintMailFragmentArgs by navArgs()
+    private val threadViewModel: ThreadViewModel by viewModels()
+    private val threadAdapter inline get() = binding.messagesList.adapter as ThreadAdapter
+
+    private var subject = ""
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt:54]: Won't this be lost to activity recreation?

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt [hunk 22]
```diff
@@ -0,0 +1,100 @@
...
+    private fun setupAdapter() = with(binding.messagesList) {
+        adapter = ThreadAdapter(
+            shouldLoadDistantResources = true,
+            isForPrinting = true,
+            onBodyWebviewFinishedLoading = {
+                CoroutineScope(Dispatchers.Main).launch {
+                    val webViewPrintAdapter = getWebViewToPrint().createPrintDocumentAdapter(subject)
+                    val printManager = requireContext().getSystemService(Context.PRINT_SERVICE) as PrintManager
+                    printManager.print(subject, webViewPrintAdapter, null)
+                }
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt:94]: Why do we need to lunch a coroutine if it's to put it on the main thread?

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt (hunk 21 above)
Comment (LunarX) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt:54]: Also do you need to save this variable here? Can't you access the value of the `threadLive?.value?.thread?.subject`?

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+
+    override fun onDestroyView() {
+        super.onDestroyView()
+        _binding = null
+    }
+
+    override fun onPause() {
+        super.onPause()
+        findNavController().popBackStack()
+    }
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt:83]: Why did you need this?

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt [hunk 24]
```diff
@@ -0,0 +1,100 @@
...
+                CoroutineScope(Dispatchers.Main).launch {
+                    val webViewPrintAdapter = getWebViewToPrint().createPrintDocumentAdapter(subject)
+                    val printManager = requireContext().getSystemService(Context.PRINT_SERVICE) as PrintManager
+                    printManager.print(subject, webViewPrintAdapter, null)
+                }
+            },
+        )
+    }
+
+    private fun getWebViewToPrint(): WebView = with(binding.messagesList[0]) { findViewById(R.id.bodyWebView) }
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt:99]: Wouldn't we want the `fullMessageWebView` instead?

File: app/src/main/res/raw/javascript_bridge.js (hunk 6 above)
Comment (LunarX) [reply app/src/main/res/raw/javascript_bridge.js:27]: Is this really where you meant that comment to be?

**Review by LunarX — CHANGES_REQUESTED** [review]

**Comment by LunarX** [comment]
Somehow you don't have the copyright update setup because they're still stuck to 2023

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt (hunk 22 above)
Comment (tevincent) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt:94]: I have this error when doing so:

```
Java.lang.RuntimeException: java.lang.Throwable: A WebView method was called on thread 'JavaBridge'. All WebView methods must be called on the same thread. (Expected Looper Looper (main, tid 2) {ed6baa7} called on Looper (JavaBridge, tid 141) {bf80207}, FYI main Looper is Looper (main, tid 2) {ed6baa7})
2024-01-30 14:48:03.788  3292-5327  System.err              com.infomaniak.mail                  W  	at android.webkit.WebView.checkThread(WebView.java:2626)
2024-01-30 14:48:03.788  3292-5327  System.err              com.infomaniak.mail                  W  	at android.webkit.WebView.createPrintDocumentAdapter(WebView.java:1151)
2024-01-30 14:48:03.788  3292-5327  System.err              com.infomaniak.mail                  W  	at com.infomaniak.mail.ui.main.thread.PrintMailFragment$setupAdapter$1$1.invoke(PrintMailFragment.kt:91)
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  	at com.infomaniak.mail.ui.main.thread.PrintMailFragment$setupAdapter$1$1.invoke(PrintMailFragment.kt:86)
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  	at com.infomaniak.mail.ui.main.thread.ThreadAdapter$bindBody$1$5.invoke(ThreadAdapter.kt:431)
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  	at com.infomaniak.mail.ui.main.thread.ThreadAdapter$bindBody$1$5.invoke(ThreadAdapter.kt:427)
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  	at com.infomaniak.mail.utils.WebViewUtils$JavascriptBridge.webviewFinishedLoading(WebViewUtils.kt:86)
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  	at android.os.MessageQueue.nativePollOnce(Native Method)
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  	at android.os.MessageQueue.next(MessageQueue.java:335)
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  	at android.os.Looper.loopOnce(Looper.java:162)
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  	at android.os.Looper.loop(Looper.java:294)
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  	at android.os.HandlerThread.run(HandlerThread.java:67)
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  Caused by: java.lang.Throwable: A WebView method was called on thread 'JavaBridge'. All WebView methods must be called on the same thread. (Expected Looper Looper (main, tid 2) {ed6baa7} called on Looper (JavaBridge, tid 141) {bf80207}, FYI main Looper is Looper (main, tid 2) {ed6baa7})
2024-01-30 14:48:03.789  3292-5327  System.err              com.infomaniak.mail                  W  	at android.webkit.WebView.checkThread(WebView.java:2621)
```

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt (hunk 23 above)
Comment (tevincent) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt:83]: To avoid going back to PrintMailFragment after printing

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt (hunk 11 above)
Comment (tevincent) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt:590]: I just tried and it seems to work fine without. I'm removing this.

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt (hunk 14 above)
Comment (tevincent) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt:147]: @KevinBoulongne let's battle with @LunarX

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt (hunk 15 above)
Comment (tevincent) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/ThreadAdapter.kt:152]: These changes are not relevant anymore. I'll revert back to the original.

File: app/src/main/java/com/infomaniak/mail/extensions/DateExtensions.kt (hunk 17 above)
Comment (LunarX) [reply app/src/main/java/com/infomaniak/mail/extensions/DateExtensions.kt]: I don't know how we want to format the date for this email but we should keep the day before the month for the most commonly used langages.

Can you recreate a format that looks like the output we would get on APIs > 25 with the french/german ordering of day and month and just spell the month in all letters?

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailViewModel.kt [hunk 25]
```diff
@@ -0,0 +1,41 @@
...
+
+    fun startPrintingService(activityContext: Context, subject: String?, webView: WebView) {
+        viewModelScope.launch(Dispatchers.Main) {
+            subject?.let { subject ->
+                val webViewPrintAdapter = webView.createPrintDocumentAdapter(subject)
+                val printManager = activityContext.getSystemService(Context.PRINT_SERVICE) as PrintManager
+                printManager.print(subject, webViewPrintAdapter, null)
+            }
+        }
+    }
```
Comment (LunarX) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailViewModel.kt:40]: To reduce the number of brackets level we usually like to write it as follow when launching coroutine like this
```suggestion
    fun startPrintingService(
        activityContext: Context,
        subject: String?,
        webView: WebView,
    ) = viewModelScope.launch(Dispatchers.Main) {
        subject?.let { subject ->
            val webViewPrintAdapter = webView.createPrintDocumentAdapter(subject)
            val printManager = activityContext.getSystemService(Context.PRINT_SERVICE) as PrintManager
            printManager.print(subject, webViewPrintAdapter, null)
        }
    }
```

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt (hunk 23 above)
Comment (LunarX) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt:70]: So leaving while it's loading gets you out as well then?

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt (hunk 24 above)
Comment (LunarX) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/PrintMailFragment.kt:88]: `fullMessageWebView` doesn't have all messages of the thread, but it has the history of other emails if it was included, I think we want to also print this

**Review by LunarX — APPROVED** [review]
